core/src/main/scala/kafka/server/FetcherEventBus.scala

```diff
@@ -0,0 +1,177 @@
+    }
+  }
+
+  /**
+   * poll() unconditionally removes the earliest item
+   * If there are no items, poll() has no effect.
+   */
+  def poll(): Unit = {
+    delayedQueue.poll()
+  }
+
+  def size = delayedQueue.size
+}
+
+/**
+ * the ConditionFactory trait is defined such that a MockCondition can be
+ * created for the purpose of testing
+ */
+trait ConditionFactory {
+  def createCondition(lock: Lock): Condition
+}
+
+object DefaultConditionFactory extends ConditionFactory {
+  override def createCondition(lock: Lock): Condition = lock.newCondition()
+}
+
+/**
+ * The FetcherEventBus supports queued events and delayed events.
+ * Queued events are inserted via the {@link #put} method, and delayed events
+ * are inserted via the {@link #schedule} method.
+ * Events are polled via the {@link #getNextEvent} method, which returns
+ * either a queued event or a scheduled event.
+ * @param time
+ */
+class FetcherEventBus(time: Time, conditionFactory: ConditionFactory = DefaultConditionFactory) {
+  private val eventLock = new ReentrantLock()
+  private val newEventCondition = conditionFactory.createCondition(eventLock)
+
+  private val queue = new PriorityQueue[QueuedFetcherEvent]
+  private val scheduler = new SimpleScheduler[DelayedFetcherEvent]
+  @volatile private var shutdownInitialized = false
+
+  def eventQueueSize() = queue.size
+
+  def scheduledEventQueueSize() = scheduler.size
+
+  /**
+   * close should be called in a thread different from the one calling getNextEvent()
+   */
+  def close(): Unit = {
+    shutdownInitialized = true
+    inLock(eventLock) {
+      newEventCondition.signalAll()
+    }
+  }
+
+  def put(event: FetcherEvent): Unit = {
+    inLock(eventLock) {
+      queue.add(new QueuedFetcherEvent(event, time.milliseconds()))
+      newEventCondition.signalAll()
+    }
+  }
+
+  def schedule(delayedEvent: DelayedFetcherEvent): Unit = {
+    inLock(eventLock) {
+      scheduler.schedule(delayedEvent)
+      newEventCondition.signalAll()
+    }
+  }
+
+  /**
+   * There are 2 cases when the getNextEvent() method is called
+   * 1. There is at least one delayed event that has become current or at least one queued event. We return either
+   * the delayed event with the earliest due time or the queued event, depending on their priority.
+   * 2. There are neither delayed events that have become current, nor queued events. We block until the earliest delayed
+   * event becomes current. A special case is that there are no delayed events at all, under which the call would block
+   * indefinitely until it is waken up by a new delayed or queued event.
+   *
+   * @return Either a QueuedFetcherEvent or a DelayedFetcherEvent that has become current. A special case is that the
+   *         FetcherEventBus is shutdown before an event can be polled, under which null will be returned.
+   */
+  def getNextEvent(): QueuedFetcherEvent = {
+    inLock(eventLock) {
+      var result : QueuedFetcherEvent = null
+
+      breakable {
```

Comment:
Does scala still implement break with an exception throw?  If so then this code might be very inefficient.

Reply:
Oh dang I didn't know about that.

Reply:
Good catch. I'll avoid using breakable here.